### ISSUE TITLE
CI documentation artifact is double-zipped

### ISSUE THREAD

**Issue by wthrowe** [sxs-collaboration/spectre#3666]
The downloadable "docs-html.zip" file contains a single file, also called "docs-html.zip", that actually contains the documentation.

**Comment by nilsvu** [comment]
This is intentional, or at least I was aware of it when I added compression to the docs artifact in #3496. It's a known issue of GitHub's `upload-artifact` action and even mentioned in their docs: https://github.com/actions/upload-artifact#zipped-artifact-downloads

**Comment by wthrowe** [comment]
At the very least, can we get the two layers to have different names?  That was extremely confusing and had me reading the documentation of unzip because I thought I was using it incorrectly.

**Comment by duetosymmetry** [comment]
I think we can skip the workflow step 'Prepare for upload' which does the first zip,
https://github.com/sxs-collaboration/spectre/blob/37590fb533eeb256d39b015ab3e270203294dd80/.github/workflows/Tests.yaml#L278-L281
because [`upload-artifact` can upload a whole directory as documented here](https://github.com/actions/upload-artifact#upload-an-entire-directory).

**Comment by nilsvu** [comment]
Last time I checked the directory wasn't automatically compressed before uploading, so it took orders of magnitude longer. That's why I added the compression step. However, I think what we should do now is `tar` the docs directory before uploading, format it correctly for GitHub Pages, then deploy it directly to GitHub Pages. It will still show up as an artifact, and named differently so this issue would be resolved. This is now possible as of https://github.blog/2022-08-10-github-pages-now-uses-actions-by-default/. It means we can delete the `gh-pages` branch and skip the custom deployment and Jekyll build steps. The feature is still in public beta, though I tried it for another project and it worked well. So should we do it now or later?

**Comment by duetosymmetry** [comment]
You're right... you still get a zip, now with an ordinary directory structure instead of another zip inside. I don't know why there is such a slowdown (I found the upload stage to take 9min30sec instead of less than 10sec).

Are you proposing to deploy docs for each PR? Or only once a PR is merged in? It's nice to be able to examine the docs as a result of the github action, but it's unclear to me which repo would be hosting the github pages with the deployed docs. If the test actions are running for the upstream repo sxs-collaboration/spectre , then is this repo's docs deployment going to get replaced every time a PR is updated? Or is there a way to get a separate deployment for each PR?

**Comment by nilsvu** [comment]
I believe the `upload-artifact` action uploads the _many_ files in the docs directory individually, like when you `scp` a directory. Only when downloading the artifact from the web UI is a zip created on demand (even when the artifact has only a single file that is already zipped, which was the original confusion that prompted this issue).

I don't propose any changes to when and where we deploy docs. FYI, we deploy docs to spectre-code.org every time a PR is merged (i.e., on every push to the develop branch). We don't currently deploy PR docs anywhere. Instead, the PR docs can be downloaded as an artifact from the GitHub Actions UI if needed. It would be nice to deploy PR docs to some server, but we haven't bothered implementing that yet.

**Comment by duetosymmetry** [comment]
That first part is a head-slapper...

Ok, so for 1. deploying the merged PRs, switch to GitHub Pages, that sounds great.

For 2. the docs from open PRs, one idea would be that a docs deployment only runs on the forked repo where the PR comes from, except this still has the problem that if you have multiple PRs from the same forked repo, then the deployed docs will keep getting thrown away by newer PRs.

Instead, the best I've experienced is with scipy's PRs (I'm sure other big projects have something similar). Their docs build actually runs on CircleCI, but I don't think that's the important part. The important part is that each PR's docs get a temporary deployment that eventually gets deleted (I don't know how long they persist). Then there is a [separate github action](https://github.com/scipy/scipy/blob/main/.github/workflows/circle_artifacts.yml) (using [larsoner/circleci-artifacts-redirector-action](https://github.com/larsoner/circleci-artifacts-redirector-action)) that takes the URL of the CircleCI artifacts, and uses the github API to [create a commit status](https://docs.github.com/en/rest/commits/statuses) with the `target_url` field set to the artifact. That way, you can just click the 'details' of the status that shows up in the PR ([example recent scipy PR](https://github.com/scipy/scipy/pull/16941)) along with all the other test statuses. Instead of taking you to a log file, it takes you to the artifact, in this case the docs.

**Comment by kidder** [comment]
@duetosymmetry FYI, we have a script (tools/pr-docs) that will locally generate documentation from a PR

**Comment by nilsvu** [comment]
For the PR docs deployment we have to choose a place to host the html files temporarily. CircleCI seems to support webserver access to its artifacts, based on what you said. GitHub Actions doesn't (yet) to my knowledge. My first thought would be https://netlify.com. Personally I don't see enough need for the feature at the moment to spend time implementing it myself, but if someone wants to do it, it would surely be nice to have.

**Comment by nilsdeppe** [comment]
Is there a path forward for closing this issue?
